### PR TITLE
Improved CLI argument handling using `picocli`

### PR DESCRIPTION
**Pull request by patilatharva**
#### Objective
Use [Picocli](https://picocli.info/) instead of Apache Commons CLI to parse command line options and parameters in Lfc.java, Lff.java and CliBase.java.

#### Reason
[Migrating from Commons CLI to picocli](https://picocli.info/migrating-from-commons-cli.html)

#### Tasks:

- [x] Refactor `CliBase.java` to implement the `Runnable` interface, to support Picocli.
- [x] Remove `runMain` in `CliBase.java` because Picocli handles simple validation and `--help`/`--version` options.
- [x] Declare options as annotated fields and remove the `CLIOption` enum (`Lfc.java`/`Lff.java`).
- [x] Implement the `run()` method as an entry point to post-argument parsing execution, in accordance with the `Runnable` interface (`Lfc.java`/`Lff.java`).
- [x] Rewrite function `filterProps` into a new function `filterPassOnProps` to create code gen properties from Picocli options (`CliBase.java`).
- [x] Refactoring for clarity: split `runTool` into `runTool` + `invokeGenerator` in `Lfc.java`, and `runTool` into `runTool` + `formatAllFiles` in `Lff.java`.
- [x] Tweak CLI tests to support Picocli output syntax (e.g. "Unrecognized option" → "Unknown option").
- [x] Ensure all tests on CI are passing.

### CODE REVIEW

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (cmnrd) [inline org.lflang/src/org/lflang/cli/Lfc.java:146]: Ye, I think this should be processed as an integer

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (cmnrd) [inline org.lflang/src/org/lflang/cli/Lfc.java:114]: Also here we might want to leave it to picocli to check if the directory exists (if possible)

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java:84]: Thanks for the suggestion! Picocli does support `java.nio.file.Path` directly - I just changed the input files parameter and the `outputPath` option to be interpreted as `Path`s instead of strings in Lfc/Lff. I've left all other options as Strings (or booleans to distinguish options that take no args) though, because later on they're passed into `LFGeneratorContext` through a `java.utils.Properties` object, which expects all option values to be strings.

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java]: As mentioned above - I realized it gets converted back to a string anyway, so I deleted the comment and left it as is.

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java]: Unfortunately Picocli doesn't do any validation beyond type conversion, so we'll still have to implement it ourselves. On the other hand, your comment gave me the idea of abstracting away redundant path parsing/validation code in `Lfc.java` and `Lff.java` into separate functions in `CliBase.java`, which made the code a lot cleaner.

**Comment by lhstrh** [comment]
Just resolved conflict to trigger a CI run. Please check if I did it correctly :-)

**Review by lhstrh — COMMENTED** [review]
This looks great, @patilatharva! I understood that you had started working on unit tests. Is that coming along or do you need help with those?

**Comment by lhstrh** [comment]
There are some strange test failures that are popping up, so this definitely needs some more work. 

You can ignore the build failure in Maven for now. Either I'll push a fix for that or we punt on it and wait until the Maven build support gets removed from this repository (we only need it to build Epoch, which will soon be hosted in its own repo).

**Comment by patilatharva** [comment]
> Just resolved conflicts to trigger a CI run. Please check if I did it correctly :-)

Thanks, looks good to me!

> There are some strange test failures that are popping up, so this definitely needs some more work.

Definitely, looking into it now.

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (cmnrd) [reply org.lflang/src/org/lflang/cli/Lfc.java:88]: Is there a fundamental reason for using `java.utils.Properties`? Maybe we should just create our own record class that passes the options in the correct types. Seems like converting to String and then back to the data type is redundant. This doesn't need to happen in this PR though, we can also leave for future work.

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (oowekyala) [inline org.lflang/src/org/lflang/cli/CliBase.java:131]: ??

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (oowekyala) [inline org.lflang/src/org/lflang/cli/Lfc.java:187]: Why protected?

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (oowekyala) [reply org.lflang/src/org/lflang/cli/Lfc.java:86]: Tbh even if the options need to be converted to a string later, it is cleaner to use the correct data types as early as possible in the data flow. It allows us to take advantage of built-in validation and improves the CLI help. Converting to a string is also very easy.

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (oowekyala) [reply org.lflang/src/org/lflang/cli/Lfc.java]: You can implement a custom validating converter very easily: https://picocli.info/#_custom_type_converters

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java:86]: That's true, I'll use the correct data types (with custom type convertors wherever needed) to make sure all option values are validated at the entry point of the CLI.

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/CliBase.java:131]: Piccoli applications are designed to be implemented with the `Runnable`/`Callable` interfaces, and I chose to make our `CliBase` class implement the `Runnable` interface, which is why it needed a `public void run()` method. `Lfc` and `Lff` required different `run()` methods, so I made the one in `CliBase` abstract and wrote custom versions for `Lfc` and `Lff`.

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java:187]: You're right, there's no reason why it shouldn't be private. I'll fix it now.

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (oowekyala) [reply org.lflang/src/org/lflang/cli/CliBase.java:131]: Hey, thanks for the patient explanation! I think a short javadoc comment here would be nice (just mentioning that PicoCli requires this so that it doesn't get refactored away). Also an Override annotation is missing.

Also I apologize that my comments are so unpolished, I intended them as notes to myself, but apparently github submitted everything when I answered the other discussion above.

**Review by revol-xut — APPROVED** [review]
Looks good to me

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/CliBase.java:131]: No worries, I agree a javadoc there would be helpful. Thanks for your comments!

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (cmnrd) [inline org.lflang/src/org/lflang/cli/CliBase.java:63]: As I understand, the FILES parameter is inherited by all cli tools. Thus, the description should be more generic.

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (cmnrd) [reply org.lflang/src/org/lflang/cli/CliBase.java:131]: Could you add the javadoc? I think it would be really helpful.

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (cmnrd) [inline org.lflang/src/org/lflang/cli/CliBase.java:157]: This appears to be specific to lfc. Why is it located in the base class?

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (cmnrd) [reply org.lflang/src/org/lflang/cli/Lfc.java:86]: Looks like you missed the option here, it is still a String and not a Path.

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (cmnrd) [inline org.lflang/src/org/lflang/cli/Lfc.java:200]: It looks wrong to me to invoke a AST transformation at this point. This should really happen in the generators. This is not an issue with this PR though, as the code already existed before.

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (cmnrd) [inline org.lflang/src/org/lflang/cli/Lfc.java:41]: This should use `LocalStrings.VERSION` to retrieve the current version.

File: org.lflang/src/org/lflang/cli/Lff.java
Comment (cmnrd) [inline org.lflang/src/org/lflang/cli/Lff.java:40]: This should use `LocalStrings.VERSION` to retrieve the current version.

**Review by cmnrd — COMMENTED** [review]
This looks like a great improvement. I added a couple of comments, mostly small things. I hope we can resolve them quickly and merge this soon!

**Review by cmnrd — CHANGES_REQUESTED** [review]
Updating review status to "request changes".

File: org.lflang/src/org/lflang/cli/Lff.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lff.java]: Done!

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java]: Done!

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/CliBase.java]: True, just fixed it.

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/CliBase.java]: Just fixed it.

File: org.lflang/src/org/lflang/cli/CliBase.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/CliBase.java:133]: Just added it!

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java]: Just fixed it! Also, I wanted to write custom type converters/validators (e.g. checking if a path actually exists), but that would require me to write a separate class implementing the `picocli.CommandLine.ITypeConverter` interface. However, our current validation logic uses protected CliBase members, like `io` and `reporter`, which I don't think I'll be able to access from the separate type converter class. Do you have any ideas for workarounds?

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/cli/Lfc.java]: This `TODO` actually needs fixing before we can merge this because otherwise the version number with be incorrect (as it now already is).

**Review by lhstrh — CHANGES_REQUESTED** [review]
Please address the version `TODO` and once fixed we can merge this. Thanks, @patilatharva!

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (lhstrh) [reply org.lflang/src/org/lflang/cli/Lfc.java]: Also, shouldn't it just read the version string (without the `lfc ` part?) Not sure though.

Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java]: It should be fixed with this commit: ([2d0662d](https://github.com/lf-lang/lingua-franca/pull/1534/commits/2d0662da45aee7c19e6e32bdc12e587df9e23f35)).

Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java]: Maybe an older version of the PR was shown in the review?

Comment (lhstrh) [reply org.lflang/src/org/lflang/cli/Lfc.java]: Yes, I was looking at an old version. Sorry for the noise! It's unclear to be what `IVersionProvider,getVersion()` is supposed to return though. Should it include the name of the command?

Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java]: True, I just added comments to make it more clear: [4768d34](https://github.com/lf-lang/lingua-franca/pull/1534/commits/4768d3405e94ba9d4bc2bbe6ba4265ecaeaeae30).

**Review by lhstrh — APPROVED** [review]

File: org.lflang/src/org/lflang/cli/Lfc.java
Comment (patilatharva) [reply org.lflang/src/org/lflang/cli/Lfc.java]: Sorry, I completely missed your point about whether we want to include the command name in the version info. To clarify, yes, based on previous tests.

**Pull request merged**
